### PR TITLE
Enable VS Test Explorer without the -vs switch

### PR DESCRIPTION
**Pull request by ViktorHofer**
Creating an intermediate runsettings file outside of the app's OutDir to enable VS Test Explorer support without the -vs switch.

I'll remove the -vs switch in the next batched rollout and for now added a warning that the switch will be removed eventually.

_Either a clean build or a libs.pretest (subset) build is necessary to consume the change locally._

cc @ericstj @danmosemsft @stephentoub @ManickaP

### CODE REVIEW

File: eng/build.ps1 [hunk 1]
```diff
@@ -61,6 +61,7 @@ if ($help -or (($null -ne $properties) -and ($properties.Contains('/help') -or $
 
 # VS Test Explorer support for libraries
 if ($vs) {
+  Write-Host "!!! VS Test Explorer now works without the -vs switch. The switch will be removed eventually. !!! "
```
Comment (safern) [inline eng/build.ps1:64]: maybe mention it will be removed on `June`? instead of `Eventually`?

File: eng/testing/runsettings.targets [hunk 2]
```diff
@@ -1,9 +1,18 @@
 <Project>
   <PropertyGroup>
-    <RunSettingsInputFilePath Condition="'$(RunSettingsInputFilePath)' == ''">$(MSBuildThisFileDirectory).runsettings</RunSettingsInputFilePath>
-    <RunSettingsOutputFilePath Condition="'$(RunSettingsOutputFilePath)' == ''">$(OutDir).runsettings</RunSettingsOutputFilePath>
+    <RunSettingsInputFilePath>$(MSBuildThisFileDirectory).runsettings</RunSettingsInputFilePath>
+    <RunSettingsIntermediateOutputFilePath>$(ArtifactsObjDir)$(TargetOS)-$(Configuration)-$(TargetArchitecture).runsettings</RunSettingsIntermediateOutputFilePath>
+    <RunSettingsAppOutputFilePath>$(OutDir).runsettings</RunSettingsAppOutputFilePath>
+
+    <CreateIntermediateRunSettingsFile Condition="'$(CreateIntermediateRunSettingsFile)' == ''">false</CreateIntermediateRunSettingsFile>
+    <RunSettingsOutputFilePath Condition="'$(CreateIntermediateRunSettingsFile)' == 'true'">$(RunSettingsIntermediateOutputFilePath)</RunSettingsOutputFilePath>
```
Comment (safern) [inline eng/testing/runsettings.targets:8]: What is the usage of the intermediate output file path?

File: eng/build.ps1 (hunk 1 above)
Comment (ViktorHofer) [reply eng/build.ps1:64]: I will definitely not write a date down here :D Consider we discover a blocker and can't remove the switch. Eventually is the best that I can do 😂

File: eng/testing/runsettings.targets (hunk 2 above)
Comment (ViktorHofer) [reply eng/testing/runsettings.targets:8]: If the app has already been built, the .runtimesettings file from the app's OutDir will be used. Otherwise the intermediate one which configuration specific will be used.

File: eng/testing/coverage.targets [hunk 3]
```diff
@@ -74,4 +74,16 @@
...
     <Exec Command="$(CoverageReportCommand)" />
   </Target>
+
+  
+  <!--
+    Clean the test results directory to guarantee that a report is generated from the
+    newest coverage results file.
+    Tracking issue https://github.com/microsoft/vstest/issues/2378.
+  -->
+  <Target Name="ClearTestResults"
```
Comment (ViktorHofer) [inline eng/testing/coverage.targets:84]: This is an unrelated change. the target was just in the wrong targets file

File: src/libraries/Directory.Build.targets [hunk 4]
```diff
@@ -136,9 +136,10 @@
   <Import Project="$(RepositoryEngineeringDir)codeOptimization.targets" />
   <Import Project="$(RepositoryEngineeringDir)references.targets" />
   <Import Project="$(RepositoryEngineeringDir)resolveContract.targets" />
-  <Import Project="$(RepositoryEngineeringDir)testing\runtimeConfiguration.targets" />
   <Import Project="$(RepositoryEngineeringDir)testing\tests.targets" Condition="'$(EnableTestSupport)' == 'true'" />
-  <Import Project="$(RepositoryEngineeringDir)testing\coverage.targets" Condition="'$(EnableCoverageSupport)' == 'true'" />
+  <Import Project="$(RepositoryEngineeringDir)testing\runtimeConfiguration.targets" />
+  <Import Project="$(RepositoryEngineeringDir)testing\runsettings.targets" Condition="'$(EnableRunSettingsSupport)' == 'true'" />
+  <Import Project="$(RepositoryEngineeringDir)testing\coverage.targets" Condition="'$(EnableRunSettingsSupport)' == 'true' or '$(EnableCoverageSupport)' == 'true'" />
```
Comment (safern) [inline src/libraries/Directory.Build.targets:142]: Whyu do we need coverage targets when we have run settings support enabled?

File: src/libraries/Directory.Build.props [hunk 5]
```diff
@@ -327,6 +327,7 @@
 
   <PropertyGroup Condition="'$(IsTestProject)' == 'true'">
     <EnableTestSupport>true</EnableTestSupport>
+    <EnableRunSettingsSupport>true</EnableRunSettingsSupport>
```
Comment (safern) [inline src/libraries/Directory.Build.props:330]: Should we set this to false when in CI, instead of conditioning the targets for that?

File: src/libraries/Directory.Build.targets (hunk 4 above)
Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets:142]: Because the .runsettings file encodes the coverages settings in it: https://github.com/dotnet/runtime/pull/36126/files#diff-5523d5d95c4ce34960f377fdfdc4ce5eR26

File: src/libraries/Directory.Build.props (hunk 5 above)
Comment (ViktorHofer) [reply src/libraries/Directory.Build.props:330]: The target below is already conditioned to not run on CI, I don't think this is necessary.

Comment (safern) [reply src/libraries/Directory.Build.props:330]: I know, just suggesting to instead disable the whole feature, hard to track when this runs because now I need to go and look at the targets conditions instead of looking at the feature flag.

Comment (ViktorHofer) [reply src/libraries/Directory.Build.props:330]: Oh actually I just saw the line that you were commenting on (I thought you were talking about that property in pretest.proj). No we shouldn't disable that flag in CI as we will use it very soon when we switch away from xunit.console.

Comment (safern) [reply src/libraries/Directory.Build.props:330]: > No we shouldn't disable that flag in CI as we will use it very soon when we switch away from xunit.console.

Hmm, then I think we should have the targets that depend on this flag, conditioned against this flag, and then disable this flag on CI for now, and whenever we enable this on CI just remove the condition on this flag.

Comment (ViktorHofer) [reply src/libraries/Directory.Build.props:330]: > Hmm, then I think we should have the targets that depend on this flag, conditioned against this flag, and then disable this flag on CI for now.

That's already the case. If the targets file isn't imported then no targets responsible for the .runsettings generation will run. I will add the condition for the sake of closing this discussion.

File: src/libraries/pretest.proj [hunk 6]
```diff
@@ -4,7 +4,9 @@
   <PropertyGroup>
     <TraversalGlobalProperties>BuildAllProjects=true</TraversalGlobalProperties>
     <NETCoreAppFrameworkIdentifier>$(TargetFrameworkIdentifier)</NETCoreAppFrameworkIdentifier>
+  </PropertyGroup>
 
+  <PropertyGroup Condition="'$(ContinuousIntegrationBuild)' != 'true'">
     <!-- Create an intermediate runsettings file to enable VSTest discovery. -->
     <EnableRunSettingsSupport>true</EnableRunSettingsSupport>
```
Comment (safern) [inline src/libraries/pretest.proj:11]: Why do we need to declare this here as well, isn't Directory.Build.props imported into this project build? That way we shouldn't need to condition this either.

File: src/libraries/Directory.Build.props (hunk 5 above)
Comment (safern) [reply src/libraries/Directory.Build.props]: Thanks, I think it is much cleaner to have one consistent condition to set a flag and then use that flag in other places than having multiple flags affecting the same feature.

File: src/libraries/pretest.proj (hunk 6 above)
Comment (ViktorHofer) [reply src/libraries/pretest.proj:11]: pretest.proj is no `$(TestProject)`: https://github.com/dotnet/runtime/pull/36126/files#diff-d3bedba39ce6da3e37ca0db83d7ef035R328

File: src/libraries/Directory.Build.props (hunk 5 above)
Comment (ViktorHofer) [reply src/libraries/Directory.Build.props]: Not arguing with that.

File: src/libraries/pretest.proj (hunk 6 above)
Comment (safern) [reply src/libraries/pretest.proj:11]: Ah, I missed that.

**Review by safern — APPROVED** [review]

**Comment by ViktorHofer** [comment]
Failure is https://github.com/dotnet/runtime/issues/29642

**Comment by krwq** [comment]
Thank you @ViktorHofer!

**Comment by GrabYourPitchforks** [comment]
Does this mean that when we run `build -s libs.pretest -rc Debug|Checked|Release`, VS will automatically pick up that runtime configuration for the purpose of running tests?

**Comment by safern** [comment]
> Does this mean that when we run build -s libs.pretest -rc Debug|Checked|Release, VS will automatically pick up that runtime configuration for the purpose of running tests?

For running the tests it should because it will use the testhost, but what I don't think it will work, will be building/intellisense of projects that have a ReferenceFromRuntime to S.P.CoreLib for example. We should figure out how to make that work.

**Comment by ViktorHofer** [comment]
> Does this mean that when we run build -s libs.pretest -rc Debug|Checked|Release, VS will automatically pick up that runtime configuration for the purpose of running tests?

It picks the runtime that's present in the matching testhost folder. Means if you build against Debug it picks the runtime that's present in the debug testhost which depends on the `-runtimeconfiguration` value that you chose when building from root.

> I don't think it will work, will be building/intellisense of projects that have a ReferenceFromRuntime to S.P.CoreLib for example. We should figure out how to make that work.

There are hacky ways to support that, like setting environment variables inside VS or opening it with an env var being set. You could temporarily set the `<RuntimeConfiguration>` property inside the project as well. @safern why do you think is the runtime's configuration for `ReferenceFromRuntime` important?

As a side note, going forward I hope we can use a P2P against the right CoreLib instead of a named reference.

**Comment by safern** [comment]
>  @safern why do you think is the runtime's configuration for ReferenceFromRuntime important?

Because of the way we resolve the reference from runtime. We translate it to a reference to System.Private.CoreLib.dll and that uses RuntimeConfiguration to resolve the location of corelib to reference it.

**Comment by ViktorHofer** [comment]
OK, so switching these `ReferenceFromRuntime` items into actual P2Ps will solve that.

**Comment by ericstj** [comment]
Are you guys over-thinking this?  I wouldn't expect CoreLib's surface area to change based on RuntimeConfiguration (though, unfortunately it does based on OS).  As such it won't really matter what S.P.C is used at build time so long as it exists.

FWIW I don't see how using a P2P is going to change anything here, except maybe force a rebuild of a different S.P.C than is in testhost.

**Comment by ViktorHofer** [comment]
> Are you guys over-thinking this? I wouldn't expect CoreLib's surface area to change based on RuntimeConfiguration (though, unfortunately it does based on OS). As such it won't really matter what S.P.C is used at build time so long as it exists.

I think @safern's point is that referenceFromRuntime resolves CoreLib based on the `RuntimeConfiguration` switch via a named reference. It doesn't build it, if it doesn't exist. Imagine you build CoreLib in Release and now try to build a library which contains a `<ReferenceFromRuntime Include="System.Private.CoreLib" />` item. The reference won't be resolved as CoreLib hasn't been built in the configured it is probed for which would be Debug.

**Comment by ericstj** [comment]
I see.  We could drop a props file as part of pretest that would indicate what CoreLib to use and use that when building in VS.  I can also imagine a P2P making this better, but I'm not sure how far out that work is.

Let's stop this discussion in merged PR.  Test the hypothetical and file an issue if something's broken.

**Comment by ViktorHofer** [comment]
Filed https://github.com/dotnet/runtime/issues/36464 to continue the discussion.
